### DISCUSSION TITLE
Fuzzy search a jsonb array

### DISCUSSION THREAD

**Discussion by saendu** [supabase/supabase#2415] [Questions] [answered]
I have a table recipes and want to do a fuzzy search in a jsonb column (ingredients) with an array but can't get my head around this. 

The recipes table has columns ``id,name,ingredients`` where ingredients is from type ::jsonb

The ``ingredients`` column contains an array with objects like so:
```
    [
        {
            "ID": 1,
            "Name": "SomeValue",
        },
        {
            "ID": 2,
            "Name": "OtherName",
        },
        ...
    ]
```

I can successfully do a exact match (case sensitive) like this:
```
const { data, error } = await supabase
        .from<any>('recipes')
        .select()
        .contains('ingredients', JSON.stringify([{Name: 'SomeValue'}]))
```
This would return the recipe, since it has an ingredient with Name: SomeValue

What I want to achieve is to find the recipe even if a user searches for 'Some'.
I tried the following approaches with no success:

- A .contains() with % or * placeholders
```
const { data, error } = await supabase
        .from<any>('recipes')
        .select()
        .contains('ingredients', JSON.stringify([{Name: 'Some%'}]))
```

- a .textSearch() with Some% and Some* or just Some  
```
returns: "operator does not exist: jsonb @@ tsquery"
```

- or .ilike() (but not supported on jsonb columns)
```
returns "operator does not exist: jsonb ~~* unknown"
```

**Comment by steve-chavez** [comment, marked as answer]
> A .contains() with % or * placeholders

Yeah, this will not work because that's a LIKE operation.

For this case I think you could create a [computed column](https://postgrest.org/en/stable/api.html#computed-columns) that aggregates the ingredients into a single string, then do the LIKE on that string.

```sql
create function ingredients_string(r recipes) returns text as $$
  select string_agg(ing->>'Name', ',') from (select jsonb_array_elements(r.ingredients) as ing) _;
$$ language sql;

-- will output: SomeValue,OtherName
```

Then you can use that computed column like any other filter on js:

```js
const { data, error } = await supabase
  .from('recipes')
  .select()
  .like('ingredients_string', 'Some%')
```

**Reply by saendu** [reply]
Thanks @steve-chavez it worked with the function. There was a small typo with the params, I had to switch type and name like so:

```sql
create or replace function ingredients_string(r recipes) returns text as $$
  select string_agg(ing->>'Name', ',') from (select jsonb_array_elements(r.ingredients) as ing) _;
$$ language sql;
```
